daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala

```diff
@@ -257,7 +264,7 @@ class Compiler private[japi] () {
    * @param tunables a map of key/value pairs, where the key is the tunable name and the value is the value to set it to
    */
   def setTunables(tunables: java.util.AbstractMap[String, String]): Unit = {
-    sCompiler.setTunables(tunables.asScala.toMap)
+    sCompiler = sCompiler.withTunables(tunables.asScala.toMap)
```

Comment:
Shoudl all these setters be marked as deprecated and we add with* methods?

Reply:
I have not yet gone so far as to modify sapi/japi, and the TDMLDFDLProcessor classes. These are one-indirection away from the actual compiler, PF, and DataProcessor objects. It's pretty easy to make them work properly with setters. 

That said, yes I do think a side-effect free design point is better, as it eliminates broad classes of otherwise subtle bugs. I just didn't want to go that far in this patch.

Reply:
It is much too involved for this patch set. Took me another whole day pulling the other threads of this review. SAPI/JAPI, and the TDMLDFDLProcessor classes all still have imperative APIs. 

They're much safer now, in that if you say, spawn lots of parse calls on varous threads, and then start calling things like setExternalDFDLVariables again on the DP object, nothing bad will happen, because those setter calls displace the state with a new state.